_source/_posts/2022-06-28-add-auth-to-any-app-with-oauth2-proxy.adoc

```diff
@@ -0,0 +1,583 @@
+        proxy_set_header X-Access-Token $token;
+        auth_request_set $auth_cookie $upstream_http_set_cookie;
+        add_header Set-Cookie $auth_cookie;
+
+        proxy_set_header Host $host;
+        proxy_set_header X-Real-IP $remote_addr;
+        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+        proxy_set_header X-Forwarded-Host $host:80;
+        proxy_set_header X-Forwarded-Port 80;
+        proxy_set_header X-Forwarded-Server $host;
+        proxy_set_header X-Forwarded-Proto http;
+
+        proxy_http_version 1.1; # <.>
+        proxy_pass http://web-app:8080/; # <.>
+    }
+}
+----
+
+<.> Don't expose this route to external clients.
+<.> Nginx makes a request to `oauth2-proxy` REST API, to verify the auth for this request.
+<.> Add headers returned from the auth request.
+<.> HTTP 1.0 is the default if not set.
+<.> Send authenticated requests to the web app.
+====
+
+Restart the docker containers and verify everything is working with HTTPie (make sure you are using port `80` now).
+
+[source,bash]
+----
+http localhost/echo "Authorization: Bearer ${TOKEN}"
+----
+
+If you remove or change the `Authorization` header, a `401` will be returned. Browser requests will now also return a `401`!
+
+Almost done! We still need to make everything work for both API clients and browsers and handle sign-out requests.
+
+== Route  all traffic through Nginx
+
+Sending all the traffic through Nginx has the added benefit of being able to control how the OAuth2 Proxy endpoints are exposed. For example, the previous section marked the `/oauth2/auth` route as "internal," so only the `auth_requst` directive can use it.
+
+In the `nginx-default.conf.template` add a couple of new `location` sections to expose the other `/oauth2` endpoints. The first `location` will handle the OAuth 2.0-related requests like the redirect callback. The second will configure the sign-out endpoint to accept only POST requests (this prevents a rogue GET request from ending the user's session).
+
+====
+[source,conf]
+----
+    location /oauth2/ {
+        proxy_pass       http://oauth2-proxy:4180; # <.>
+        proxy_set_header Host                    $host;
+        proxy_set_header X-Real-IP               $remote_addr;
+        proxy_set_header X-Scheme                $scheme;
+    }
+
+    location = /oauth2/sign_out { # <.>
+        # Sign-out mutates the session, only allow POST requests
+        if ($request_method != POST) {
+            return 405;
+        }
+
+        proxy_pass       http://oauth2-proxy:4180;
+        proxy_set_header Host                    $host;
+        proxy_set_header X-Real-IP               $remote_addr;
+        proxy_set_header X-Scheme                $scheme;
+    }
+...
+----
+
+<.> Send the OAuth callback and logout requests to oauth2-proxy.
+<.> Only allow POST requests to the sign-out endpoint.
+====
+
+NOTE: The sign-out endpoint does not use a CSRF token. TODO: link to Alisa's post on this topic.
```

Comment:
This one? https://developer.okta.com/blog/2022/07/08/spa-web-security-csrf-xss#validate-requests-for-authenticity-to-mitigate-csrf

Reply:
```suggestion
NOTE: OAuth2 Proxy's sign-out endpoint does not use a link:/blog/2022/07/08/spa-web-security-csrf-xss#validate-requests-for-authenticity-to-mitigate-csrf[CSRF token]. 
```

Should this link be more obvious like:
> ... does not use a CSRF token, link:/blog/2022/07/08/spa-web-security-csrf-xss#validate-requests-for-authenticity-to-mitigate-csrf[learn about why that's important]